### ISSUE TITLE
Need Chat Persistent across across all the browser

### ISSUE THREAD

**Issue by Bradrajkumar** [microsoft/BotFramework-WebChat#3156]
Hi All,
hope you all are doing good.
i want chat persistent feature on react webchat as well as simple webchat.
the feature can be described as:- 
1. user opened the webchat on browser(either react or simple webchat version).
2. user did some conversation.
3. in between the conversation flow user opened a new tab in browser and load the same webchat(in this scenario the conversation happened in 1st tab should show as well as user should be able to continue the chat on current tab).
4. now user opened the new window with same webchat(in this scenario also the webchat should able to  persist the chat in previous tab or window and should be able to continue the chat).

exception:- 
------------
suppose user is in dialog flow in this scenario how we can handle the same across all the browser tabs or windows.

these are few of the scenario i thought of and atleast wanted to  execute the same.
please suggest some more ideas on this and let me know if in any ways we can make it possible.

@compulim @tdurnford 
your comments will be truly appreciated. :-)

**Comment by compulim** [comment]
(I am not the expert in this area, it's on the service side. But I kind of know how it should work.)

tl;dr as long as the token retains the same conversation ID, Web Chat (or Direct Line channel, to be specific), will restore the conversation history, up to ~23 hours.

To survive <kbd>F5</kbd> or any types of page navigation:

- On page load after navigation (including <kbd>F5</kbd>), the browser should resend request for another token
- The newly generated token should retain the same conversation ID
   - I am unsure which API to use, @ckkashyap and @swagatmishra2007 do you know how to generate a token for resuming existing conversation, e.g. after token expired?
- When the token with same conversation ID is loaded in Direct Line channel (regardless of Web Chat), the conversation history will be restored automatically

Tips on handling the token:

- General
   - Do not save the token in the browser, because it will expire in ~60 minutes
   - Do not save conversation history in the browser, because it is not trivial to clear personal-identifiable information (PII) from browser
- Security perspective
   - The token should be treated as secure as other auth token, i.e. with anti-forgery measurement
   - The token is a key to PII and SSO, i.e. leaking the token equals to leaking PII and SSO access to other systems, which could have very serious consequence
   - Never allow anyone to generate token with any conversation ID
   - Additional read at https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-direct-line-3-0-authentication?view=azure-bot-service-4.0#azure-bot-service-authentication

**Comment by Bradrajkumar** [comment]
@compulim  thanks for your comment. please forward this to your service team to have a look once, i am eagerly waiting for correct and perfect approach.

**Comment by compulim** [comment]
Yes, I pinged both @ckkashyap and @swagatmishra2007 in another thread. 😉

**Comment by Bradrajkumar** [comment]
hi @compulim ,

i haven't heard from @ckkashyap and @swagatmishra2007 yet. will you please help me out to have a conversation or please share me the thread where you have pinged both of them so that i will be able to do one-to-one communication.

thanks

**Comment by ckkashyap** [comment]
Hey @Bradrajkumar ,
It is not possible to generate a new token with a conversation ID that was in an expired token.
Ideally, you need to host a token server which generates a token  in exchange for a secret. Please let us know if you need more information on setting up a token server.
Regards,
Kashyap

**Comment by ckkashyap** [comment]
Hey @Bradrajkumar ,
Please let us know if we can perhaps have a call set up so that we could understand your use case better and offer suggestions that could work for you.
Regards,
Kashyap

**Comment by srinaath** [comment]
@Bradrajkumar Could you respond to this ticket. If not we will be closing it due to inactivity.
